csi/src/dev/iscsi.rs

```diff
@@ -0,0 +1,249 @@
+            "invalid target specification: {}",
+            path
+        )))
+    }
+}
+
+pub(super) type IscsiAttach = IscsiDevice;
+
+impl TryFrom<&Url> for IscsiDevice {
+    type Error = DeviceError;
+
+    fn try_from(url: &Url) -> Result<Self, Self::Error> {
+        let host = url
+            .host_str()
+            .ok_or_else(|| DeviceError::new("missing host"))?;
+
+        let segments: Vec<&str> = url
+            .path_segments()
+            .ok_or_else(|| DeviceError::new("no path segment"))?
+            .collect();
+
+        if segments.is_empty()
+            || (segments.len() == 1 && segments[0].is_empty())
+        {
+            return Err(DeviceError::new("no path segment"));
+        }
+
+        if segments.len() > 2 {
+            return Err(DeviceError::new("too many path segments"));
+        }
+
+        let lun: u16 = if segments.len() > 1 {
+            segments[1].parse()?
+        } else {
+            0
+        };
+
+        let components: Vec<&str> = segments[0].split(':').collect();
+
+        if components.len() != 2 {
+            return Err(DeviceError::new("invalid IQN"));
+        }
+
+        let uuid = extract_uuid(components[1]).map_err(|error| {
+            DeviceError::from(format!("invalid UUID: {}", error))
+        })?;
+
+        let portal =
+            format!("{}:{}", host.to_string(), url.port().unwrap_or(3260));
+
+        Ok(IscsiDevice::new(portal, segments[0].to_string(), uuid, lun))
+    }
+}
+
+#[tonic::async_trait]
+impl Attach for IscsiAttach {
+    async fn attach(&self) -> Result<(), DeviceError> {
+        match IscsiAdmin::find_session(&self.portal, &self.iqn) {
+            Ok(found) => {
+                if found {
+                    // session already exists - nothing to do
+                    return Ok(());
+                }
+            }
+            Err(error) => {
+                return Err(DeviceError::from(format!(
+                    "iscsiadm command (session) failed: {}",
+                    error
+                )));
+            }
+        }
+
+        if let Err(error) = IscsiAdmin::discover(&self.portal, &self.iqn) {
+            return Err(DeviceError::from(format!(
+                "iscsiadm command (discovery) failed: {}",
+                error
+            )));
+        }
+
+        if let Err(error) = IscsiAdmin::login(&self.portal, &self.iqn) {
+            let _ = IscsiAdmin::delete(&self.portal, &self.iqn);
+            return Err(DeviceError::from(format!(
+                "iscsiadm command (login) failed: {}",
+                error
+            )));
+        }
+
+        Ok(())
+    }
+
+    async fn find(&self) -> Result<Option<DeviceName>, DeviceError> {
+        let path = self.to_path();
+
+        let mut enumerator = Enumerator::new()?;
+
+        enumerator.match_subsystem("block")?;
+        enumerator.match_property("DEVTYPE", "disk")?;
+
+        for device in enumerator.scan_devices()? {
```

Comment:
Here the same thing applies, you practically know what you are looking for we can get rid of a lot of if's here.

Reply:
as mentioned i dont fully understand why we have .lookup() and .find() the both find devices... i get that this one is "iscsi specific" but the code is the same as in lookup. IOW it looks to me that .lookup can find your device if its a nvme OR iscsi device?

Reply:
When we call lookup() it's to find a device to detach. At that point we only have a UUID. We don't know if its iscsi/nvme or anything else and so we look for all possibilities in 1 pass. In the other case, when we use find(), we already have a device (derived from the URI) and we call find() after the attach() to find out what device gets assigned to it. Since we know what type of device it is already, we don't need to bother checking for the other possibilities which we know can't happen. Yes lookup() should work in that case, but at least half the comparisons are unnecessary.

Reply:
Have refactored this avoid duplicating common code. As an added bonus, it is also hopefully now much clearer to see what we are matching against in udev.